agent/src/config/config.cpp

```diff
@@ -17,205 +17,203 @@
+    bool found = false;
+
+    for (const auto &dep : m_dependencies) {
+        UniqueKey config = dep.first;
         std::string group = std::get<0>(config);
         std::string key = std::get<1>(config);
-        ConfigType type = std::get<2>(config);
-
-        log_debug() << "Checking for presence of mandatory config: \"" << key << "\"";
-        /* TODO: The below only checks for presence, it should also add the values to something
-         *       that can be used for retrieval later. */
-        try {
-            switch (type) {
-                case ConfigType::String:
-                    getStringValue(group, key);
-                    break;
-                case ConfigType::Integer:
-                    getIntegerValue(group, key);
-                    break;
-                case ConfigType::Boolean:
-                    getBooleanValue(group, key);
-                    break;
-                default:
-                    break;
+
+        // Check if the dependee exists in the configs we got from the sources
+        found = (std::find(m_allConfigs.begin(), m_allConfigs.end(), config) != m_allConfigs.end());
+
+        // If a dependee was present among the configs, all its dependencies needs to be present as well
+        if (found) {
+            if (allDepsSatisfied(dep.second)) {
+                log_debug() << "All dependencies found for config " << group << "::" << key;
+            } else {
+                // It is a fatal error if one or more dependencies were not found
+                std::string message = "One or more dependencies were not met for config " + group + "::" + key;
+                log_error() << message;
+                throw ConfigDependencyError(message);
             }
-        } catch (ConfigError &error) {
-            log_error() << "Missing mandatory config: \"" << key << "\"";
-            throw ConfigMandatoryError("Missing mandatory config: \"" + key + "\"");
         }
     }
 }
 
-void Config::loadConfig()
+bool Config::allDepsSatisfied(const std::vector<UniqueKey> &dependencies) const
 {
-    try {
-        m_config = m_loader->loadConfig();
-    } catch (Glib::Error &error) {
-        throw ConfigError();
+    for (const UniqueKey &dependency : dependencies) {
+        bool found = (std::find(m_allConfigs.begin(), m_allConfigs.end(), dependency) != m_allConfigs.end());
+
+        if (!found) {
+            return false;
+        }
     }
-}
 
-std::string Config::getStringValue(const std::string &group, const std::string &key) const
-{
-    return getValue<std::string>(group, key, m_stringOptions);
+    return true;
 }
 
-int Config::getIntegerValue(const std::string &group, const std::string &key) const
+void Config::assertAllMandatoryPresent()
 {
-    return getValue<int>(group, key, m_intOptions);
-}
+    // There might not be any configs which are mandatory
+    if (m_mandatory.empty()) {
+        return;
```

Comment:
This check is not needed as you iterate over the list. I.e. if `m_mandatory` is empty then the for loop below will never be entered, hence no exception can be thrown.